homeassistant/components/egardia.py

```diff
@@ -0,0 +1,139 @@
+    'HOME': STATE_ALARM_ARMED_HOME,
+    'TRIGGERED': STATE_ALARM_TRIGGERED,
+    'UNKNOWN': STATE_UNKNOWN,
+}
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_HOST): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Optional(CONF_VERSION, default=DEFAULT_VERSION): cv.string,
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+        vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+        vol.Optional(CONF_REPORT_SERVER_CODES): vol.All(cv.ensure_list),
+        vol.Optional(CONF_REPORT_SERVER_ENABLED,
+                     default=DEFAULT_REPORT_SERVER_ENABLED): cv.boolean,
+        vol.Optional(CONF_REPORT_SERVER_PORT,
+                     default=DEFAULT_REPORT_SERVER_PORT): cv.port,
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+
+def setup(hass, config):
+    """Set up the Egardia platform."""
+    from pythonegardia import egardiadevice
+    from pythonegardia import egardiaserver
+
+    name = config[DOMAIN].get(CONF_NAME)
+    hass.data[D_EGARDIANM] = name
+    username = config[DOMAIN].get(CONF_USERNAME)
+    password = config[DOMAIN].get(CONF_PASSWORD)
+    host = config[DOMAIN].get(CONF_HOST)
+    port = config[DOMAIN].get(CONF_PORT)
+    rs_enabled = config[DOMAIN].get(CONF_REPORT_SERVER_ENABLED)
+    hass.data[D_EGARDIARSENABLED] = rs_enabled
+    rs_port = config[DOMAIN].get(CONF_REPORT_SERVER_PORT)
+    rs_codes = config[DOMAIN].get(CONF_REPORT_SERVER_CODES)
+    hass.data[D_EGARDIARSCODES] = rs_codes
+    version = config[DOMAIN].get(CONF_VERSION)
+    try:
+        hass.data[D_EGARDIASYS] = egardiadevice.EgardiaDevice(
+            host, port, username, password, '', version)
+    except requests.exceptions.RequestException:
+        raise exc.PlatformNotReady()
+    except egardiadevice.UnauthorizedError:
+        _LOGGER.error("Unable to authorize. Wrong password or username")
+        return
+
+    # add egardia alarm device
+    def alarm_loaded(event, data=None):
+        """Check if egardia platform has loaded."""
+        if event is DOMAIN:
+            # configure egardia server, including callback
+            if rs_enabled:
+                # Set up the egardia server
+                _LOGGER.info("Setting up EgardiaServer")
+                try:
+                    if D_EGARDIASRV not in hass.data:
+                        server = egardiaserver.EgardiaServer('', rs_port)
+                        bound = server.bind()
+                        if not bound:
+                            raise IOError("Binding error occurred while " +
+                                          "starting EgardiaServer")
+                        hass.data[D_EGARDIASRV] = server
+                        server.start()
+                except IOError:
+                    return
+                hass.data[D_EGARDIASRV].register_callback(
+                    hass.data[D_EGARDIADEV].handle_status_event)
```

Comment:
This should be done in the entity that needs the callback. Do it in `async_added_to_hass` method.

Reply:
Not sure which entity that should be.

Reply:
Eg the `EgardiaAlarm` entity that is saved under this key `D_EGARDIADEV`. But it should be done for each entity that needs a callback to update state.

Reply:
Ok, thanks. I moved this piece of code to `alarm_control_panel\egardia.py` in the `__init__` function. Let me know if that is OK.

Reply:
No that's not good since the entity will not have been added to home assistant at that point. Callbacks for entities should generally be registered in `async_added_to_hass`.